src/Kernel/Magic/DebugMagic.cs

```diff
@@ -0,0 +1,224 @@
+        ///      The symbol resolver used by this magic command to find
+        ///      operations or functions to be simulated.
+        /// </summary>
+        public ISymbolResolver SymbolResolver { get; }
+
+        /// <summary>
+        ///     The configuration source used by this magic command to control
+        ///     simulation options (e.g.: dump formatting options).
+        /// </summary>
+        public IConfigurationSource ConfigurationSource { get; }
+
+        /// <inheritdoc />
+        public override ExecutionResult Run(string input, IChannel channel) =>
+            RunAsync(input, channel).Result;
+
+        private async Task HandleStartMessage(Message message) => await Task.Run(() =>
+        {
+            var content = (message.Content as UnknownContent);
+            var session = content?.Data?["debug_session"];
+            if (session == null)
+            {
+                // TODO: Communicate error back here!
+            }
+            else
+            {
+                shellServer.SendShellMessage(
+                    new Message
+                    {
+                        Header = new MessageHeader
+                        {
+                            MessageType = "iqsharp_debug_reply"
+                        },
+                        Content = new UnknownContent
+                        {
+                            Data = new Dictionary<string, object>
+                            {
+                                ["debug_session"] = session
+                            }
+                        }
+                    }
+                    .AsReplyTo(message)
+                );
+            }
+        });
+
+        private async Task HandleAdvanceMessage(Message message) => await Task.Run(() =>
+        {
+            logger.LogDebug("Got debug advance message:", message);
+            var content = (message.Content as UnknownContent);
+            var session = content?.Data?["debug_session"];
+            if (session == null)
+            {
+                logger.LogWarning("Got debug advance message, but debug_session was null.", message);
+            }
+            else
+            {
+                var sessionGuid = Guid.Parse(session.ToString());
+                ManualResetEvent? @event = null;
+                lock (sessionAdvanceEvents)
+                {
+                    @event = sessionAdvanceEvents[sessionGuid];
+                }
+                @event.Set();
+            }
+        });
+
+        private async Task WaitForAdvance(Guid session)
+        {
+            await Task.Run(() =>
+            {
+                ManualResetEvent? @event = null;
+                // Find the event we need to wait on.
+                lock (sessionAdvanceEvents)
+                {
+                    @event = sessionAdvanceEvents[session];
+                }
+                @event.Reset();
+                @event.WaitOne();
+            }, cancellationTokenSource.Token);
```

Comment:
What's the purpose of this cancellation token? I don't think I see it being used as well to cancel this `Run` method.

Reply:
Actually this raises a good point. It would be awesome if the `%debug` magic could integrate with Jupyter kernel interrupt, so that the debug session is cancelled when the user chooses the "interrupt kernel" command in Jupyter.

To do this, you would simply rename the `Run` method in this class to `RunCancellable`, which also contains a `CancellationToken` parameter:
https://github.com/microsoft/iqsharp/blob/03db172008c671e193c93d45c0097ed2eccc34f2/src/Jupyter/Magic/AbstractMagic.cs#L153-L163
And then you'd pass that token into `RunAsync` and eventually through to here. You'd then also want to pass that token into the `Task.Run(() => symbol.Operation.RunAsync(qsim, inputParameters))` statement later on so that the whole operation is cancelled.

This is a separate change and can certainly be done in a later PR; just wanted to make you aware of this.

For now, as @theRoughCode points out, since this is unused you could probably just remove it.

Reply:
I will add this to my list of issues that I need to address before we merge the final feature branch to master